### DISCUSSION TITLE
@nx/gradle and composite builds

### DISCUSSION THREAD

**Discussion by juan-abril** [nrwl/nx#23095] [Feature Requests]
#  @nx/gradle and composite builds
Hi Team, good morning, In `Gradle`, there are two ways to structure a project: `multi-module` project or `composite builds`. I know the plugin is currently working with `multi-module` projects, but I would like to ask for support for `composite builds`.

## Not ask for root build.gradle
I think you are asking for the root `build.gradle` file to include the `project-report` plugin in all Gradle projects. However, sometimes the monorepo has different languages, not just Java or Gradle. Maybe we could add an option to the plugin to not ask for the root build.gradle and leave the responsibility of adding the `project-report` plugin to the developer.

## Discover Gradle projects dynamically
I think the plugin is currently asking for the root `build.gradle` file to discover the Gradle projects. However, in a monorepo, the projects are not always in the root `build.gradle` file. I suggest adding a way to discover the Gradle projects dynamically. For example, the plugin already has a way to [get all](https://github.com/nrwl/nx/blob/master/packages/gradle/src/plugin/dependencies.ts#L22) the `build.gradle` files in the workspace. Maybe we could use that to discover the Gradle projects.

**Example of workspace**
```
.
├── apps/
│   └── api/
│       ├── build.gralde
│       └── settings.gradle
├── libs/
│   ├── base-utils/
│   │   ├── build.gralde
│   │   └── settings.gradle
│   ├── string-utils/
│   │   ├── build.gralde
│   │   └── settings.gradle
│   └── number-utils/
│       ├── build.gralde
│       └── settings.gradle
└── tools/
    └── greeting-plugin/
        ├── build.gralde
        └── settings.gradle
```
**Dependency Tree**

```mermaid
  graph TD;
      api-->string-utils;
      api-->number-utils;
      string-utils-->base-utils;
      number-utils-->greeting-plugin;
```

## Composite Build
I noticed that for `multi-module` you are asking for the [following pattern](https://github.com/nrwl/nx/blob/master/packages/gradle/src/plugin/dependencies.ts#L109"): `dep.startsWith('project ')` and `+--- project :number-utils`. In composite build, the dependency report has a different pattern. For example: `+--- com.example:number-utils -> project :number-utils`. I suggest adding another if statement to support that.

Thank guys for the plugin.

**Comment by jaysoo** [comment]
Thanks for the feedback. We'll look into this.

**Comment by xiongemi** [comment]
this issue has been fixed. please see the example here: https://github.com/xiongemi/nx-gradle-composite-build
it should be fixed from 19.4.0. 

there is a migration scirpt packages/gradle/src/migrations/19-4-0/add-project-report-all.ts

if you want to migrate manually, in every build file, add below code
if using kotlin, in build.gradle.kts:
```
tasks.register("projectReportAll") {
    // All project reports of subprojects
    allprojects.forEach {
        dependsOn(it.tasks.get("projectReport"))
    }

    // All projectReportAll of included builds
    gradle.includedBuilds.forEach {
        dependsOn(it.task(":projectReportAll"))
    }
}
```

if using grovy, in build.gradle:
```
tasks.register("projectReportAll") {
        // All project reports of subprojects
        allprojects.forEach {
            dependsOn(it.tasks.getAt("projectReport"))
        }
    
        // All projectReportAll of included builds
        gradle.includedBuilds.forEach {
            dependsOn(it.task(":projectReportAll"))
        }
    }
```
